doc/crypto/api/ops/kdf.rst

```diff
@@ -153,6 +153,98 @@ Key derivation algorithms
+
+        | `PSA_KEY_TYPE_HMAC` (for the secret key)
+        | `PSA_KEY_TYPE_DERIVE` (for the secret key)
+        | `PSA_KEY_TYPE_RAW_DATA` (for the other inputs)
+
+.. macro:: PSA_ALG_SP800_108_COUNTER_CMAC
+    :definition: ((psa_algorithm_t)0x08000800)
+
+    .. summary::
+        Macro to build a NIST SP 800-108 conformant, counter-mode KDF algorithm based on CMAC.
+
+    This is the CMAC-based, counter mode key derivation function specified by :cite-title:`SP800-108`, §4.1.
+
+    This key derivation algorithm uses the following inputs, in any order:
+
+    *   `PSA_KEY_DERIVATION_INPUT_SECRET` is the secret input keying material, *K*\ :sub:`IN`. This must be a block-cipher key that is compatible with the CMAC algorithm. See also `PSA_ALG_CMAC`, and must be input using `psa_key_derivation_input_key()`.
+    *   `PSA_KEY_DERIVATION_INPUT_LABEL` is the *Label*. It is optional; if omitted, the *Label* is a zero-length string.
+    *   `PSA_KEY_DERIVATION_INPUT_CONTEXT` is the *Context*. It is optional; if omitted, the *Context* is a zero-length string.
+
+    Each input may only be passed once.
+
+    This algorithm uses the output length as part of the derivation process. In the derivation this value is *L*, the required output size in bits. After setup, the initial capacity of the key derivation operation is 2\ :sup:`29` - 1 bytes (``0x1fffffff``). The capacity can be set to a lower value by calling `psa_key_derivation_set_capacity()`.
+
+    When the first output is requested, the value of *L* is calculated as *L* = :code:`8 * psa_key_derivation_get_capacity()`. Subsequent calls to `psa_key_derivation_set_capacity()` are not permitted for this algorithm.
+
+    The derivation is constructed as described in :cite:`SP800-108` §4.1, with the following details:
+
+    *   The iteration counter *i* and output length *L* are encoded as big-endian, 32-bit values.
+    *   The mitigation to make the CMAC-based construction robust is implemented.
+
+    The resulting output stream *K*\ :sub:`1` || *K*\ :sub:`2` || *K*\ :sub:`3` || ... is computed as:
+
+        | *K*\ :sub:`0` = CMAC( *K*\ :sub:`IN`, *Label* || ``0x00`` || *Context* || [ *L* ]\ :sub:`4` )
+        | *K*\ :sub:`i` = CMAC( *K*\ :sub:`IN`, [ *i* ]\ :sub:`4` || *Label* || ``0x00`` || *Context* || [ *L* ]\ :sub:`4` || *K*\ :sub:`0` ), for *i* = 1, 2, 3, ...
+
+    Where [ *x* ]\ :sub:`n` is the big-endian, *n*-byte encoding of the integer *x*.
+
+    .. rationale::
+
+        SP 800-108 describes a set of general constructions for key derivation algorithms, with flexibility for specific implementation requirements.
+
+        The precise definition provided here enables compatibility between different implementations of the |API|.
+
+    .. subsection:: Compatible key types
+
+        | `PSA_KEY_TYPE_AES` (for the secret key)
+        | `PSA_KEY_TYPE_ARIA` (for the secret key)
+        | `PSA_KEY_TYPE_CAMELLIA` (for the secret key)
+        | `PSA_KEY_TYPE_SM4` (for the secret key)
```

Comment:
Do we have a reason to avoid having PSA_KEY_TYPE_DERIVE here?

Reply:
In addition the key must have a size that is valid for the block-cipher, when used with CMAC.

To ensure that this interoperates well with APIs that can produce keys, such as `psa_pake_get_shared_key()`, we need to permit those APIs to return block-cipher keys, where the API generates unbiased pseudorandom output, and the output size matches a valid key size. See #114.

Reply:
Understood, thank you for the explanation!